radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+    for change in changes {
+        let substate_id = change.id;
+        by_re_node_module
+            .entry(ReNodeModule::new(substate_id.0, substate_id.1))
+            .or_insert_with(|| Vec::new())
+            .push(IdChange::new(substate_id.2, change.changed));
+    }
+    by_re_node_module
+}
+
+struct TreeRoot<P> {
+    hash: Hash,
+    node: TreeNode<P>,
+}
+
+fn put_substate_changes<S: TreeStore>(
+    store: &mut S,
+    current_version: Option<Version>,
+    re_node_module: &ReNodeModule,
+    changes: Vec<IdChange<SubstateOffset, Hash>>,
+) -> Option<TreeRoot<SubstateOffset>> {
+    let substate_entry = get_substate_entry(store, current_version, re_node_module);
+    let (substate_version, substate_root) = break_tuple(substate_entry);
+    let mut substate_tree_store = NestedTreeStore::new(store, re_node_module, substate_root);
+    let substate_root_hash = put_changes(
+        &mut substate_tree_store,
+        substate_version,
+        current_version.unwrap_or(0) + 1,
+        changes
+            .into_iter()
+            .map(|change| to_substate_change(change))
+            .collect(),
+    );
+    let substate_root_node = substate_tree_store
+        .extract_new_root()
+        .map_payload(|payload| match payload {
+            Payload::ReNodeModule(_) => panic!("unexpected payload in substate layer"),
+            Payload::Substate(substate_offset) => substate_offset,
+        });
+    if matches!(substate_root_node, TreeNode::Null) {
+        None
+    } else {
+        Some(TreeRoot {
+            hash: substate_root_hash,
+            node: substate_root_node,
         })
-        .collect();
+    }
+}
+
+fn get_substate_entry<S: ReadableTreeStore>(
+    store: &S,
+    current_version: Option<Version>,
+    re_node_module: &ReNodeModule,
+) -> Option<(Version, TreeNode<SubstateOffset>)> {
+    current_version
+        .and_then(|version| {
+            JellyfishMerkleTree::new(store)
+                .get_with_proof(hash(scrypto_encode(re_node_module).unwrap()), version)
+                .unwrap()
+                .0
+        })
+        .map(|(_, (payload, version))| {
+            let substates_root = match payload {
+                Payload::ReNodeModule(payload) => payload.substates_root,
+                Payload::Substate(_) => panic!("unexpected payload in RE node module layer"),
+            };
+            (version, substates_root)
+        })
+}
+
+struct LeafChange {
+    key_hash: Hash,
+    new_payload: Option<(Hash, Payload)>,
+}
+
+fn put_changes<S: TreeStore>(
+    store: &mut S,
+    current_version: Option<Version>,
+    next_version: Version,
+    changes: Vec<LeafChange>,
+) -> Hash {
     let (root_hash, update_result) = JellyfishMerkleTree::new(store)
         .batch_put_value_set(
-            value_set
+            changes
                 .iter()
-                .map(|(x, y)| (x.clone(), y.as_ref()))
+                .map(|change| (change.key_hash.clone(), change.new_payload.as_ref()))
```

Comment:
If you use `into_iter()` in the line above, you should be able to avoid the clone.

Reply:
Or is the reference required for the new payload?

Reply:
the reference is required by the ultra-weird API of that low-level, opensource JMT method: `batch_put_value_set(..., value_set: Vec<(Hash, Option<&(Hash, K)>)>, ...)`.
Maybe they had a reason, maybe their `K` can weight megabytes - I don't know and I am not touching that :p
If it helps visually, that `.clone()` call can actually be skipped :P (and maybe the compiler optimizes it away?)